docs/development/testguide.rst

```diff
@@ -665,6 +665,32 @@ a real-world example::
    so the `astropy.tests.helper.catch_warnings` context manager is
    preferred.
 
+Testing exceptions
+==================
+
+Just like the handling of warnings described above, tests that are
+designed to trigger certain errors should verify that an exception of
+the expected type is raised in the expeced place.  This is efficiently
+done by running the tested code inside the
+`pytests.raises <https://docs.pytest.org/en/latest/assert.html#assertraises>`_
+context manager.  Its optional ``matches`` argument allows to check the
+error message for any patterns using ``regex`` syntax.  For example the
+matches ``pytest.raises(OSError, match=r'^No such file')`` and
+``pytest.raises(OSError, match=r'or directory$')`` would be equivalent
+to ``assert str(err).startswith(No such file)`` and ``assert
+str(err).endswith(or directory)``, respectively, on the raised error
+message ``err``.
+For matching multi-line messages you need to pass the ``(?s)``
+`flag <https://docs.python.org/3/library/re.html#re.S>`_
```

Comment:
use intersphinx here, too

Reply:
> use intersphinx here, too

What about all the existing hard links? I'm willing to work on that in principle, but haven't got around to install the host of sphinx-astropy extensions locally, so I am working a bit in the blind for now. The `pytest.raises` link e.g. does not refer to the reference docs, but to the tutorial page, so it might need some experimenting to get that resolved by intersphinx.

Reply:
We've tried to get rid of most. If you see any from packages that we intersphinx map, or that we should rather intersphinx, changing them would be great.
We do have the ones from sphinx-astropy, and add a few here in our config: https://github.com/astropy/astropy/blob/master/docs/conf.py#L69

Reply:
(in the intersphinx maps everything that is anchored in the package docs page, looking at the link at the anchor gives the clue what to link)

Reply:
Is there another syntax to refer to the intersphinx map besides 
> :ref:\`ref name \<package:reftarget\>\` ?

Because for that I could only find a single instance (for packagetemplate, so not really external). But not a great lot of hard links for mapped packages either - almost all of the pytest references in `testguide.rst` and a bunch of pandas links.

Reply:
What exactly did you mean by "looking at the link at the anchor gives the clue what to link"?
What I've been able to extract from the Sphinx docs so far was that 
```
python -m sphinx.ext.intersphinx https://docs.pytest.org/en/stable/objects.inv
```
would list the human-readable inventory of the pytest docs, which does contain a link to 
```
	pytest helpers                           Pytest API and builtin fixtures         : builtin.html#pytest-helpers
```
and then per analogy to
> :ref:\`comparison manual \<python:comparisons\>\`

I should use the label in the first "column" to substitute the old reference to `<https://pytest.org/en/latest/builtin.html#pytest-helpers>`.

So, any idea what the warning
```
/home/circleci/project/docs/development/testguide.rst:627:py:obj reference target not found: pytest:pytest helpers
```
is complaining about?
Obviously using a label with blank space is not the most helpful, but the docs don't give any clue whether this should be escaped with extra quotes or anything.

But I also noticed that the `#pytest-helpers` label does not seem to link to any particular section on the web page, and the suggested redirection to `Command Reference` actually links the `distutils` documentation! So maybe it's the pytest docs that have gone somewhat stale.

Reply:
Giving up on the "pytest helper functions" reference - I could not identify a single label with a valid target on the "reference" doc page that did not raise some warning like
```
undefined label: pytest:pytest.raises (if the link has no caption the label must precede a section header)
```
and broke the docs build.

Reply:
I'm sorry for not getting back sooner. All the changes now looks very nice, thanks for digging up the references. I didn't know about the `python -m sphinx.ext.intersphinx` either, I used to dig up the mapping manually by decoding the inv file, which is not the least painless procedure (but then found empirically that the anchors are somewhat usable for this purpose, and I was apparently wrong with that, so I'm sorry about that.)